### PR TITLE
Iterator refactoring.

### PR DESCRIPTION
**Pull request by feuerste**
This speeds up node iterator generation for already completed octrees and simplifies the OctreeDataProvider trait by removing the `num_points` method.

### CODE REVIEW

File: src/octree/mod.rs [hunk 1]
```diff
@@ -96,8 +95,7 @@ pub struct Octree {
 }
 
 pub struct PointsInBoxIterator<'a> {
-    octree_data_provider: &'a OctreeDataProvider,
-    octree_meta: &'a OctreeMeta,
+    octree: &'a Octree,
```
[the remark this answers is not in the document]
Comment (feuerste) [reply src/octree/mod.rs]: I still need them, without the compiler complains.

File: src/octree/mod.rs [hunk 2]
```diff
@@ -358,8 +357,7 @@ impl Octree {
             }
         }
         PointsInBoxIterator {
-            octree_data_provider: &*self.data_provider,
-            octree_meta: &self.meta,
+            octree: &*self,
```
[the remark this answers is not in the document]
Comment (feuerste) [reply src/octree/mod.rs]: You are right, `&self` is sufficient here. I adjusted it.

File: src/octree/node.rs [hunk 3]
```diff
@@ -279,6 +279,7 @@ impl NodeIterator {
         octree_data_provider: &OctreeDataProvider,
```
[the remark this answers is not in the document]
Comment (feuerste) [reply src/octree/node.rs]: Done.

**Comment by nnmm** [comment]
Not the assigned reviewer, but LGTM!

**Comment by feuerste** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user feuerste. Merge queue now has a length of 2.

**Pull request merged**